Description: disable Ubuntu overlay scrollbars as they don’t play well with Qt
Author: Aurélien Gâteau <aurelien.gateau@canonical.com>
Last-Update: 2013-04-22
Forwarded: not-needed

--- a/src/widgets/styles/qgtkstyle_p.cpp
+++ b/src/widgets/styles/qgtkstyle_p.cpp
@@ -99,6 +99,7 @@ Q_GLOBAL_STATIC(QGtkStyleUpdateScheduler, styleScheduler)
 
 Ptr_gtk_container_forall QGtkStylePrivate::gtk_container_forall = 0;
 Ptr_gtk_init QGtkStylePrivate::gtk_init = 0;
+Ptr_ubuntu_gtk_set_use_overlay_scrollbar QGtkStylePrivate::ubuntu_gtk_set_use_overlay_scrollbar = 0;
 Ptr_gtk_style_attach QGtkStylePrivate::gtk_style_attach = 0;
 Ptr_gtk_window_new QGtkStylePrivate::gtk_window_new = 0;
 Ptr_gtk_widget_destroy QGtkStylePrivate::gtk_widget_destroy = 0;
@@ -351,6 +352,7 @@ void QGtkStylePrivate::resolveGtk() const
     QLibrary libgtk(QLS("gtk-x11-2.0"), 0, 0);
 
     gtk_init = (Ptr_gtk_init)libgtk.resolve("gtk_init");
+    ubuntu_gtk_set_use_overlay_scrollbar = (Ptr_ubuntu_gtk_set_use_overlay_scrollbar)libgtk.resolve("ubuntu_gtk_set_use_overlay_scrollbar");
     gtk_window_new = (Ptr_gtk_window_new)libgtk.resolve("gtk_window_new");
     gtk_style_attach = (Ptr_gtk_style_attach)libgtk.resolve("gtk_style_attach");
     gtk_widget_destroy = (Ptr_gtk_widget_destroy)libgtk.resolve("gtk_widget_destroy");
@@ -549,6 +551,8 @@ void QGtkStylePrivate::initGtkWidgets() const
         QGtkStylePrivate::gtk_init (NULL, NULL);
 #ifndef Q_OS_MAC
         XSetErrorHandler(qt_x_errhandler);
+        if (QGtkStylePrivate::ubuntu_gtk_set_use_overlay_scrollbar)
+            QGtkStylePrivate::ubuntu_gtk_set_use_overlay_scrollbar(false);
 #endif
 
         // make a window
--- a/src/widgets/styles/qgtkstyle_p_p.h
+++ b/src/widgets/styles/qgtkstyle_p_p.h
@@ -129,6 +129,7 @@ typedef char* (*Ptr_gconf_client_get_string)(GConfClient*, const char*, GError *
 typedef bool (*Ptr_gconf_client_get_bool)(GConfClient*, const char*, GError **);
 
 typedef void (*Ptr_gtk_init)(int *, char ***);
+typedef void (*Ptr_ubuntu_gtk_set_use_overlay_scrollbar) (gboolean);
 typedef GtkWidget* (*Ptr_gtk_window_new) (GtkWindowType);
 typedef GtkStyle* (*Ptr_gtk_style_attach)(GtkStyle *, GdkWindow *);
 typedef void (*Ptr_gtk_widget_destroy) (GtkWidget *);
@@ -344,6 +345,7 @@ public:
 
     static Ptr_gtk_container_forall gtk_container_forall;
     static Ptr_gtk_init gtk_init;
+    static Ptr_ubuntu_gtk_set_use_overlay_scrollbar ubuntu_gtk_set_use_overlay_scrollbar;
     static Ptr_gtk_style_attach gtk_style_attach;
     static Ptr_gtk_window_new gtk_window_new;
     static Ptr_gtk_widget_destroy gtk_widget_destroy;
